scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,669 @@
+                        'trying new random parameters')]
+                    raise ValueError(self._message[0])
+                self._x = self._lower + self._random_state.random_sample(
+                        self._x.size) * (self._upper - self._lower)
+                reinit_counter += 1
+            else:
+                init_error = False
+
+    def _visiting(self, step):
+        """
+        Assignement of components values based on visiting distribution.
+        The way of exploring space depends on the Markov chain stepping
+        """
+        # It it is the first part of the markov chain
+        # Changing all components at the same time
+        if step < self._x.size:
+            visits = np.array([self._visita() for _ in range(self._x.size)])
+            visits[visits > 1.e8] = 1.e8 * self._random_state.random_sample()
+            visits[visits < -1e8] = -1.e8 * self._random_state.random_sample()
+            self._x = visits + self._xbackup
+            a = self._x - self._lower
+            b = np.fmod(a, self._xrange) + self._xrange
+            self._x = np.fmod(b, self._xrange) + self._lower
+            self._x[np.fabs(self._x - self._lower) < 1.e-10] += 1.e-10
+        else:
+            # Second part of the markov chain
+            # Now change only one component at a time
+            visit = self._visita()
+            if visit > 1.e8:
+                visit = 1.e8 * self._random_state.random_sample()
+            elif visit < -1e8:
+                visit = -1.e8 * self._random_state.random_sample()
+            index = step - self._x.size
+            self._x[index] = visit + self._xbackup[index]
+            a = self._x[index] - self._lower[index]
+            b = np.fmod(
+                a, self._xrange[index]) + self._xrange[index]
+            self._x[index] = np.fmod(
+                b, self._xrange[index]) + self._lower[index]
+            if np.fabs(self._x[index] - self._lower[
+                    index]) < 1.e-10:
+                self._x[index] += 1.e-10
+
+    def _run_markov_chain(self, step):
+        for j in range(self.markov_length):
+            if j == 0:
+                self._emin_unchanged = True
+            if step == 0 and j == 0:
+                self._emin_unchanged = False
+            # Keeping old location before visiting new place
+            self._xbackup = np.array(self._x)
+            self._visiting(j)
+            self._energy(self._x)
+            if self._etot < self._etot0:
+                # We get a better energy value
+                self._etot0 = np.array(self._etot)
+                if self._etot < self._emin:
+                    self._emin = np.array(self._etot)
+                    self._xmin = np.array(self._x)
+                    self._emin_unchanged = False
+                    self._index_no_emin_update = 0
+            else:
+                # We do not have improvement but do we accept location
```

Comment:
Add a question mark to the comment.  I misread it over and over again as "we *do* accept location".  :)

Also, this whole if else block could be in a separate function.